### PR TITLE
Fix nil deploymentkey crash

### PR DESCRIPTION
**Pull request by ryuyu**
Fixes a crash that was caused by attempting to report status when the deployment key wasn't specified in config.xml.

Addresses https://github.com/Microsoft/cordova-plugin-code-push/issues/108

@lostintangent @geof90

### CODE REVIEW

File: src/ios/CodePush.m [hunk 1]
```diff
@@ -58,10 +58,14 @@ - (void)handleUnconfirmedInstall:(BOOL)navigate {
 }
 
 - (void)updateSuccess:(CDVInvokedUrlCommand *)command {
+    NSString* deploymentKey = [command argumentAtIndex:0 withDefault:nil andClass:[NSString class]];
+    if ([deploymentKey length] == 0) {
```
Comment (dlebu) [inline src/ios/CodePush.m]: Is this `nil` proof?

File: typings/codePush.d.ts [hunk 2]
```diff
@@ -187,10 +187,11 @@ interface CodePushCordovaPlugin {
      * Calling this function is required on the first run after an update. On every subsequent application run, calling this function is a noop.
      * If using sync API, calling this function is not required since sync calls it internally.
      *
+     * @param deploymentKey Optional string deployment key for this package.
      * @param notifySucceeded Optional callback invoked if the plugin was successfully notified.
      * @param notifyFailed Optional callback invoked in case of an error during notifying the plugin.
      */
-    notifyApplicationReady(notifySucceeded?: SuccessCallback<void>, notifyFailed?: ErrorCallback): void;
+    notifyApplicationReady(deploymentKey?: string, notifySucceeded?: SuccessCallback<void>, notifyFailed?: ErrorCallback): void;
```
Comment (dlebu) [inline typings/codePush.d.ts]: This is an API breaking change. Is this really necessary? Can't we be smarter in the native code and remember the custom deployment key?

File: www/codePush.ts [hunk 3]
```diff
@@ -43,11 +43,12 @@ class CodePush implements CodePushCordovaPlugin {
      * Calling this function is required on the first run after an update. On every subsequent application run, calling this function is a noop.
      * If using sync API, calling this function is not required since sync calls it internally.
      *
+     * @param deploymentKey Optional string deployment key for this package.
      * @param notifySucceeded Optional callback invoked if the plugin was successfully notified.
      * @param notifyFailed Optional callback invoked in case of an error during notifying the plugin.
      */
-    public notifyApplicationReady(notifySucceeded?: SuccessCallback<void>, notifyFailed?: ErrorCallback): void {
-        cordova.exec(notifySucceeded, notifyFailed, "CodePush", "updateSuccess", []);
+    public notifyApplicationReady(deploymentKey?: string, notifySucceeded?: SuccessCallback<void>, notifyFailed?: ErrorCallback): void {
```
Comment (dlebu) [inline www/codePush.ts]: JS source file should be checked in as well.

File: typings/codePush.d.ts (hunk 2 above)
Comment (scottbommarito) [inline typings/codePush.d.ts]: I don't mind adding deploymentKey to notifyApplicationReady, but it needs to be the last argument. Not only does it break the API if it's at the front, but it's more consistent with checkForUpdate's custom deployment key specification if it's at the back.

File: src/ios/CodePush.m (hunk 1 above)
Comment (ryuyu) [inline src/ios/CodePush.m]: Yes. Calling length on nil will return 0.

File: typings/codePush.d.ts (hunk 2 above)
Comment (ryuyu) [inline typings/codePush.d.ts]: Yes I forgot to consider that. I will just move it to the end.

Comment (dlebu) [inline typings/codePush.d.ts]: I think, as much as possible, we should keep our interfaces for react and cordova in sync. AFAIK, the react plugin does not expose this parameter. @scottbommarito We should not do it if it is not needed.

File: src/android/CodePush.java [hunk 4]
```diff
@@ -103,13 +103,24 @@ protected Void doInBackground(Void... params) {
...
-    private boolean execUpdateSuccess(CallbackContext callbackContext) {
+    private boolean execUpdateSuccess(CordovaArgs args, CallbackContext callbackContext) {
+        String deploymentKey = "";
+        try {
+            deploymentKey = args.getString(0);
+        } catch (JSONException e) {
+            // no-op becuase we don't care if we didn't get a deployment key from script; we will just try to use the one from preferences
+        }
+
+        if (deploymentKey.equals(null) || deploymentKey.equals("") || deploymentKey.equals("null") || deploymentKey.equals("undefined")) {
```
Comment (geof90) [inline src/android/CodePush.java]: `deploymentKey.equals(null)` will never be true. If `deploymentKey == null`,  `deploymentKey.equals(null)` will throw a `NullPointerException`.

Comment (ryuyu) [inline src/android/CodePush.java]: Right. Forogt about that. Fixed.

**Comment by scottbommarito** [comment]
imo I think this problem is a little more complex than this solution.

basically, we now have 3 separate deployment keys
1 - deployment key in config.xml
2 - deployment key passed to checkForUpdate
3 - "default" deployment key (used when 1 and 2 are unspecified)

right now, the default key is null. when we don't have a key in config.xml and no key was passed into checkForUpdate, then we crash.

I agree with @dlebu 's assessment that we can be smarter in our native code to prevent the default key from being null. I like adding the deployment key as an argument to notifyApplicationReady, but I think that if the deployment key is unspecified, it should use the key used in the last call to checkForUpdate (when you call notifyApplicationReady you're probably trying to validate the last update u checked for). If not, it should use the key in the config.xml.

File: src/android/CodePush.java [hunk 5]
```diff
@@ -310,13 +321,6 @@ private void handleAppStart() {
                             this.codePushPackageManager.clearFailedUpdates();
                             this.codePushPackageManager.clearPendingInstall();
                             this.codePushPackageManager.clearInstallNeedsConfirmation();
-                            try {
-                                String appVersion = Utilities.getAppVersionName(cordova.getActivity());
-                                codePushReportingManager.reportStatus(CodePushReportingManager.Status.STORE_VERSION, null, appVersion, mainWebView.getPreferences().getString(DEPLOYMENT_KEY_PREFERENCE, null), this.mainWebView);
-                            } catch (PackageManager.NameNotFoundException e) {
-                                // Should not happen unless the appVersion is not specified, in which case we can't report anything anyway.
-                                e.printStackTrace();
-                            }
```
Comment (geof90) [inline src/android/CodePush.java]: With this code removed, wouldn't that mean we wont be able to report binary updates?

File: www/codePush.ts [hunk 6]
```diff
@@ -293,7 +294,7 @@ class CodePush implements CodePushCordovaPlugin {
             CodePushUtil.copyUnassignedMembers(defaultOptions, syncOptions);
         }
 
-        window.codePush.notifyApplicationReady();
+        window.codePush.notifyApplicationReady(null, null, syncOptions.deploymentKey);
```
Comment (geof90) [inline www/codePush.ts]: Can't we make the deploymentKey the first param? Then we wouldn't have to do `null, null`. Seems more idiomatic for node style async callback functions too.

Comment (geof90) [inline www/codePush.ts]: Sorry, just saw that @dlebu commented on this previously.

**Comment by ryuyu** [comment]
Pulling this PR in favor of minimum change required to unblock.

**Pull request closed without merging**